ui/apps/platform/src/Containers/Integrations/IntegrationForm/Forms/CloudSourceIntegrations/OcmIntegrationForm.tsx

```diff
@@ -0,0 +1,412 @@
+
+function testTokenValue(value, context: yup.TestContext): boolean {
+    const requireSecretField =
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        !!context?.from[2]?.value?.updateCredentials;
+    const clientIdEntered =
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        !!context?.from[2]?.value?.cloudSource?.credentials?.clientId?.trim();
+    const clientSecretEntered =
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        !!context?.from[2]?.value?.cloudSource?.credentials?.clientSecret?.trim();
+
+    if (!requireSecretField || clientIdEntered || clientSecretEntered) {
+        return true;
+    }
+    return !!value?.trim();
+}
+
+function testClientValue(value, context: yup.TestContext): boolean {
+    const requireSecretField =
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        !!context?.from[2]?.value?.updateCredentials;
+    const tokenEntered =
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        !!context?.from[2]?.value?.cloudSource?.credentials?.secret?.trim();
+
+    if (!requireSecretField || tokenEntered) {
+        return true;
+    }
+    return !!value?.trim();
+}
+
+export const validationSchema = yup.object().shape({
+    cloudSource: yup.object().shape({
+        name: yup.string().trim().required('Integration name is required'),
+        type: yup.string().matches(/TYPE_OCM/),
+        credentials: yup.object().shape({
+            secret: yup.string().test('secret-test', 'Token is required', testTokenValue),
+            clientId: yup.string().test('client-id-test', 'Client ID is required', testClientValue),
+            clientSecret: yup
+                .string()
+                .test('client-secret-test', 'Client secret is required', testClientValue),
+        }),
+        ocm: yup.object().shape({
+            endpoint: yup.string().trim().required('Endpoint is required'),
+        }),
+        skipTestIntegration: yup.bool(),
+    }),
+    updatePassword: yup.bool(),
+});
+
+export type CloudSourceIntegrationFormValues = {
+    cloudSource: CloudSourceIntegration;
+    updateCredentials: boolean;
+};
+export const defaultValues: CloudSourceIntegrationFormValues = {
+    cloudSource: {
+        id: '',
+        name: '',
+        type: 'TYPE_OCM',
+        credentials: {
+            secret: '',
+            clientId: '',
+            clientSecret: '',
+        },
+        skipTestIntegration: true,
+        ocm: {
+            endpoint: 'https://api.openshift.com',
+        },
+    },
+    updateCredentials: true,
+};
+
+function OcmIntegrationForm({
+    initialValues = null,
+    isEditable = false,
+}: IntegrationFormProps<CloudSourceIntegration>): ReactElement {
+    const formInitialValues = { ...defaultValues, ...initialValues };
```

Comment:
Did you mean to spread `...initialValues` into here as well? It looks like all of the fields on that object match the fields on `defaultValues.cloudSource`, which are manually set directly below if `initialValues` is not `null`, so would be applied to the object at the top level as well as at the `cloudSource` level.

Also small nit: I tend to use lodash's `cloneDeep` for things like initializing a default, as the spread operator on `defaultValues` only does a shallow copy, and over time code changes can lead to issues where the module level `defaultValues` is actually corrupted by property writes in the component. (Or in this futuristic year of 2024 we can use the native `structuredClone()` to do the same 😎 )